### ISSUE TITLE
[Feature] Implement `Response.dispose` akin to `APIResponse.dispose`.

### ISSUE THREAD

**Issue by michaeleveringham** [microsoft/playwright-python#2191]
**Let us know what functionality you'd like to see in Playwright and what your use case is.**

Currently we can handle caching on a per-response level for `APIResponse` objects, but not for `Response` objects. It seems that the underlying object can be disposed using a private method; `Response._impl_obj._dispose()` deletes and removes the object from `BrowserContext._impl_obj.objects`. However I'm not sure how wise or effective it is to use this directly.

Example:
```
import asyncio

from playwright.async_api import async_playwright


async def main():
    playwright = await async_playwright().start()
    browser = await playwright.chromium.launch()
    context = await browser.new_context()
    page = await context.new_page()

    response = await context.request.get("https://github.com")
    # APIResponse, works as expected.
    await response.dispose()

    response = await page.goto("https://github.com")
    
    # Would raise an AttributeError.
    # await response.dispose()

    # Works, but unclear how safe. Mapped object still exists, too.
    response._impl_obj._dispose()

asyncio.run(main())
```

**Do you think others might benefit from this as well?**

Yes. This gives consistency to the models across the project and is especially helpful when handling intercepted responses.

**Comment by mxschmitt** [comment]
As of today we dispose the objects automatically if they exceed a certain number. Before adding a new API it would be relevant for us to understand your use-case better, are you re-using the same context over and over again? Are you running into memory leaks?

When you close the context, we dispose all the objects too.

**Comment by michaeleveringham** [comment]
You're correct, I am re-using the context. I have some cases where I re-create the context but with lengthy usecases (e.g. testing long-pattern behaviours that involve a large number of requests) as this can cause memory to become an issue quickly in Kubernetes, but it's a bit tedious to preserve and re-create the context after each request. 

It also can be difficult when the number of requests is not static or not known; e.g. if we are handling requests via a route handler, we cannot dispose these directly and so the only recourse (re-create context) is a potential issue as it'll dump all requests, even those that we have not finished handling yet. 
Example:
```python
browser = page.context.browser

async def recreate_context()
    # Save the context desired attributes first.
    await browser.context.close()
    # Re-create with savedattributes.
    await browser.new_context()

async def handler(request):
    response = await request.response()
    # *May be unreachable if it's been disposed when context was closed elsewhere in the loop.*
    body = await response.body()
    await recreate_context()

await browser.on("requestfinished", handler)
```

I have seen this potentially related issue, but wasnt sure if it would be entirely relevant here. https://github.com/microsoft/playwright/issues/6319#issuecomment-1736438896

**Comment by mxschmitt** [comment]
Yes its relevant, we closed https://github.com/microsoft/playwright/issues/6319#issuecomment-1736438896 after we limited the handles per type, so it should leak anymore.

I'm curious if its maybe possible for you to share a repro which leaks.

**Comment by michaeleveringham** [comment]
I'll try to craft a reproduction here and report back.

**Comment by michaeleveringham** [comment]
Okay, I think I have a pretty solid example here:

```python
import asyncio
import random

from playwright.async_api import async_playwright


COUNT = 0

BLOCKED_TYPES = [
    "image",
    "font",
    "stylesheet",
]


async def counter():
    global COUNT
    COUNT += 1
    print(f"Request #{COUNT}")


async def route_handler(route, request):
    if route.request.resource_type in BLOCKED_TYPES:
        await route.abort()
    else:
        await route.continue_()


async def request_handler(request):
    await counter()
    return request


async def setup():
    playwright = await async_playwright().start()
    browser = await playwright.firefox.launch()
    context = await browser.new_context(
        java_script_enabled=True,
    )
    page = await context.new_page()
    # Try to conserve memory by blocking most content types.
    await page.route("**/*", route_handler)
    return page


async def cycle_pages():
    """Randomly browse XHR-heavy pages."""
    urls = [
        "https://www.pinterest.com/ideas/",
        "https://youtube.com/",
        "https://www.cnn.com/",
    ]
    page = await setup()
    page.on("requestfinished", request_handler)
    while True:
        await page.goto(random.choice(urls))
        await asyncio.sleep(1)


async def scroll_forever():
    """Scroll down an XHR-heavy infinite scroll site."""
    page = await setup()
    # Add route handler after navigation so we don't dispose main page.
    await page.goto("https://reddit.com")
    page.on("requestfinished", request_handler)
    while True:
        print("Scrolling down...")
        await page.mouse.wheel(0, 10000)


asyncio.run(cycle_pages())
#asyncio.run(scroll_forever())
```

These examples both browse while blocking most "large" content types, and just print the total request count after each finished request. The first (`cycle_pages`) just hits 3 different XHR-heavy sites in a loop; the second (`scroll_forever`) scrolls a single XHR-heavy page. After about 2000 requests, I'm over 2GB of memory (from a starting point of about 350MB).

I may be oversimplifying or making assumptions, but my thought is that if the finished XHR requests could be disposed, memory usage would go down. For context, if I scroll a similar number of posts in Chrome manually, the memory profiler shows the tab at under 300MB.

Or perhaps being able to limit the total memory allocation, or number of total objects in the cache, would be a better workaround? Though not entirely related as the above would give more control over disposing of select responses.

I hope this helps, please let me know if I can provide further information.

**Comment by mxschmitt** [comment]
Which Playwright version are you using? Since a few versions we have an improved memory management logic implemented which should make it much better.

If you look at Memory, which process are looking at? Chromium, Python, or Node?

**Comment by michaeleveringham** [comment]
Confirmed I am using `playwright==1.40.0`. I'm just looking at memory locally for this, see attached photo. But I first noticed this in Kubernetes when I hit some `OOMKilled` with memory max set to 6GB.
![image](https://github.com/microsoft/playwright-python/assets/36551116/eaec3c4b-aa3e-4dd2-a04a-90a4cfb881ba)
